RecoHGCal/TICL/plugins/TrackstersMergeProducerV3.cc

```diff
@@ -0,0 +1,776 @@
+    tracksterTile.fill(tracksterIteration, t.barycenter().eta(), t.barycenter().phi(), tracksterId);
+    LogDebug("TrackstersMergeProducerV3") << "Adding tracksterId: " << tracksterId << " into bin [eta,phi]: [ "
+                                          << tracksterTile[tracksterIteration].etaBin(t.barycenter().eta()) << ", "
+                                          << tracksterTile[tracksterIteration].phiBin(t.barycenter().phi())
+                                          << "] for iteration: " << tracksterIteration << std::endl;
+
+    tracksterId++;
+  }
+}
+
+void TrackstersMergeProducerV3::dumpTrackster(const Trackster &t) const {
+  auto e_over_h = (t.raw_em_pt() / ((t.raw_pt() - t.raw_em_pt()) != 0. ? (t.raw_pt() - t.raw_em_pt()) : 1.));
+  LogDebug("TrackstersMergeProducerV3")
+      << "\nTrackster raw_pt: " << t.raw_pt() << " raw_em_pt: " << t.raw_em_pt() << " eoh: " << e_over_h
+      << " barycenter: " << t.barycenter() << " eta,phi (baricenter): " << t.barycenter().eta() << ", "
+      << t.barycenter().phi() << " eta,phi (eigen): " << t.eigenvectors(0).eta() << ", " << t.eigenvectors(0).phi()
+      << " pt(eigen): " << std::sqrt(t.eigenvectors(0).Unit().perp2()) * t.raw_energy() << " seedID: " << t.seedID()
+      << " seedIndex: " << t.seedIndex() << " size: " << t.vertices().size() << " average usage: "
+      << (std::accumulate(std::begin(t.vertex_multiplicity()), std::end(t.vertex_multiplicity()), 0.) /
+          (float)t.vertex_multiplicity().size())
+      << " raw_energy: " << t.raw_energy() << " regressed energy: " << t.regressed_energy()
+      << " probs(ga/e/mu/np/cp/nh/am/unk): ";
+  for (auto const &p : t.id_probabilities()) {
+    LogDebug("TrackstersMergeProducerV3") << std::fixed << p << " ";
+  }
+  LogDebug("TrackstersMergeProducerV3") << " sigmas: ";
+  for (auto const &s : t.sigmas()) {
+    LogDebug("TrackstersMergeProducerV3") << s << " ";
+  }
+  LogDebug("TrackstersMergeProducerV3") << std::endl;
+}
+
+void TrackstersMergeProducerV3::produce(edm::Event &evt, const edm::EventSetup &es) {
+  edm::ESHandle<CaloGeometry> geom = es.getHandle(geometry_token_);
+  rhtools_.setGeometry(*geom);
+  auto resultTrackstersMerged = std::make_unique<std::vector<Trackster>>();
+  auto resultCandidates = std::make_unique<std::vector<TICLCandidate>>();
+
+  tfSession_ = es.getData(tfDnnToken_).getSession();
+
+  TICLTracksterTiles tracksterTile;
+  std::vector<bool> usedTrackstersMerged;
+  std::vector<int> indexInMergedCollTRKEM;
+  std::vector<int> indexInMergedCollEM;
+  std::vector<int> indexInMergedCollTRK;
+  std::vector<int> indexInMergedCollHAD;
+  std::vector<bool> usedSeeds;
+
+  // associating seed to the index of the trackster in the merged collection and the iteration that found it
+  std::map<int, std::vector<std::pair<int, TracksterIterIndex>>> seedToTracksterAssociator;
+  edm::Handle<std::vector<reco::Track>> track_h;
+  evt.getByToken(tracks_token_, track_h);
+  const auto &tracks = *track_h;
```

Comment:
```suggestion
  const auto &tracks = evt.get(tracks_token_);
```
here and below. (I think this comment got lost and was not implemented from earlier)

Reply:
this is also done except for the cases where we require the `edm::Handle`s later to build the `TICLCandidates`